### PR TITLE
perform app validation on the app copy instead of the app itself when creating a build

### PR DESCRIPTION
**Pull request by snopoke**
## Product Description
Improve performance of creating a build.

## Technical Summary
Part 2 of https://dimagi-dev.atlassian.net/browse/USH-1225

Prevent re-generating app build files when creating a build. The build files are currently generated during validation which is the first step when creating a build and then again when creating the build files for the actual build doc.

This PR moves the validation setup so that it runs on the 'build' doc instead of the main app doc.

Best review by commit.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage
App build is already covered by tests. Also additional tests have been added.

### QA Plan
I have tested this locally as follows:
* make a build
* make a build with errors
* app preview with working app and with an app with errors
* web apps
* push linked app with build
* build linked app

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/app_manager/tasks.py [hunk 1]
```diff
@@ -44,12 +45,7 @@ def autogenerate_build(app, username):
 
 @serial_task('{app_id}-{version}', max_retries=0, timeout=60 * 60)
 def autogenerate_build_task(app_id, domain, version, comment):
-    app = get_app(domain, app_id)
-    errors = app.validate_app()
-    if not errors:
-        copy = app.make_build(comment=comment)
-        copy.is_auto_generated = True
-        copy.save(increment_version=False)
+    make_async_build_v2(app_id, domain, version, allow_prune=True, comment=comment)
```
Comment (esoergel) [inline corehq/apps/app_manager/tasks.py:48]: so `autogenerate_build_task` and `make_async_build_v2` are essentially identical?  Can we remove one of them?  (In a separate PR, perhaps)

File: corehq/apps/app_manager/tests/test_apps.py [hunk 2]
```diff
@@ -275,17 +276,23 @@ def testBuildTemplateApps(self, mock):
     @patch('corehq.apps.app_manager.models.validate_xform', return_value=None)
     def testGetLatestBuild(self, mock):
         factory = AppFactory(build_version='2.40.0')
-        factory.new_basic_module('register', 'case', with_form=False)
-        factory.app.save()
+        m0, f0 = factory.new_basic_module('register', 'case')
+        f0.source = BLANK_TEMPLATE.format(xmlns=f0.unique_id)
```
Comment (esoergel) [inline corehq/apps/app_manager/tests/test_apps.py]: Any reason this shouldn't be part of `new_basic_module`?

File: corehq/apps/app_manager/tests/util.py [hunk 3]
```diff
@@ -199,3 +200,34 @@ def delete_all_apps():
         )
         for row in res:
             Application.get_db().delete_doc(row['doc'])
+
+
+def get_simple_form(xmlns=None):
```
Comment (esoergel) [inline corehq/apps/app_manager/tests/util.py:205]: :+1:

**Review by esoergel — APPROVED** [review]
Really excited to see the performance improvement from this.  Even with that aside, I like the one-step build process - much nicer than calling validate and build with minor differences in a bunch of places.

**Comment by snopoke** [comment]
Build is failing because of caching and mocking in other tests. I'll look into that more tomorrow.

File: corehq/apps/app_manager/tasks.py (hunk 1 above)
Comment (snopoke) [reply corehq/apps/app_manager/tasks.py:48]: mostly likely - I didn't want to do that in this PR

File: corehq/apps/app_manager/tests/test_apps.py (hunk 2 above)
Comment (snopoke) [reply corehq/apps/app_manager/tests/test_apps.py]: It causes a lot of test failures

Comment (esoergel) [reply corehq/apps/app_manager/tests/test_apps.py]: fair enough

File: corehq/apps/app_manager/models.py [hunk 4]
```diff
@@ -4521,6 +4523,10 @@ def make_build(self, comment=None, user_id=None):
             # but check explicitly so as not to change the _id if it exists
             copy._id = uuid.uuid4().hex
 
+        errors = copy.validate_app()
+        if errors:
+            raise AppValidationError(errors)
```
Comment (orangejenny) [inline corehq/apps/app_manager/models.py:4528]: Is there any kind of cleanup to do here? What prevents `copy` from showing up on the releases page?

**Review by orangejenny — COMMENTED** [review]
Haven't read closely, but this approach makes a ton of sense.

This doesn't have any affect on the validation that happens on the form/module settings pages whenever you load them, does it? I don't think it would, but looked around briefly and didn't find where that validation actually happens, so I wanted to check.

File: corehq/apps/app_manager/models.py (hunk 4 above)
Comment (esoergel) [reply corehq/apps/app_manager/models.py:4528]: `copy` is actually not saved until later, outside `make_build`.

**Comment by snopoke** [comment]
> This doesn't have any affect on the validation that happens on the form/module settings pages whenever you load them, does it? I don't think it would, but looked around briefly and didn't find where that validation actually happens, so I wanted to check.

Those aren't impacted. They call `validate_module_for_build` and `validate_form_for_build` which run the validation directly. This only impacts places that call `app.make_build`

**Comment by snopoke** [comment]
@orangejenny @esoergel I've rolled back some of the test changes and taken a different approach that avoids having to patch 90% of the app manager tests. The new approach is to set the build spec in `AppFactory` instead of in `Application.new_app`. This means I can set a 'test' spec rather than having to call the real method get an actual spec which hits the DB.

Apologies for the force push.

**Comment by snopoke** [comment]
@esoergel build finally passing on this

File: corehq/apps/app_manager/tests/test_apps.py [hunk 5]
```diff
@@ -26,18 +26,18 @@
...
 from corehq.apps.app_manager.views.apps import load_app_from_slug
 from corehq.apps.app_manager.views.utils import update_linked_app
 from corehq.apps.builds.models import BuildSpec
 from corehq.apps.domain.shortcuts import create_domain
 from corehq.apps.linked_domain.applications import link_app
-from corehq.apps.linked_domain.models import DomainLink
 from corehq.apps.userreports.tests.utils import get_sample_report_config
 
 
+@patch_validate_xform()
```
Comment (esoergel) [inline corehq/apps/app_manager/tests/test_apps.py:40]: :+1:

File: corehq/apps/linked_domain/tests/test_linked_apps.py [hunk 6]
```diff
@@ -78,7 +71,22 @@ def tearDownClass(cls):
         cls.domain_link.delete()
         cls.domain_obj.delete()
         cls.linked_domain_obj.delete()
-        super(BaseLinkedAppsTest, cls).tearDownClass()
+        super().tearDownClass()
+
+
+@flag_enabled('CAUTIOUS_MULTIMEDIA')
+@patch_validate_xform()
+class BaseLinkedAppsTest(BaseLinkedDomainTest):
```
Comment (esoergel) [inline corehq/apps/linked_domain/tests/test_linked_apps.py:79]: > (saves about 20s on the test suite)

:100:

**Review by esoergel — APPROVED** [review]

**Pull request merged**